pgactivity/types.py

```diff
@@ -685,28 +762,82 @@ def default(
         cls,
         *,
         memory: Optional[MemoryInfo] = None,
-        swap: Optional[MemoryInfo] = None,
+        swap: Optional[SwapInfo] = None,
         load: Optional[LoadAverage] = None,
     ) -> "SystemInfo":
         """Zero-value builder.
 
         >>> SystemInfo.default()  # doctest: +NORMALIZE_WHITESPACE
-        SystemInfo(memory=MemoryInfo(percent=0.0, used=0, total=0),
-                   swap=MemoryInfo(percent=0.0, used=0, total=0),
+        SystemInfo(memory=MemoryInfo(used=0, buff_cached=0, free=0, total=0),
+                   swap=SwapInfo(used=0, free=0, total=0),
                    load=LoadAverage(avg1=0.0, avg5=0.0, avg15=0.0),
                    io_read=IOCounter(count=0, bytes=0, chars=0),
-                   io_write=IOCounter(count=0, bytes=0, chars=0),
-                   max_iops=0)
+                   io_write=IOCounter(count=0, bytes=0, chars=0), max_iops=0)
         """
         return cls(
             memory or MemoryInfo.default(),
-            swap or MemoryInfo.default(),
+            swap or SwapInfo.default(),
             load or LoadAverage.default(),
             IOCounter.default(),
             IOCounter.default(),
+            0,
         )
 
 
+@attr.s(frozen=True, auto_attribs=True, slots=True)
+class ServerInformation:
+    # Fetched from the database
+    xact_count: int
+    insert: int
+    update: int
+    delete: int
+    tuples_returned: int
+    total_size: int
+    cache_hit_ratio: Pct = attr.ib(converter=Pct)
+    max_dbname_length: int
+    uptime: timedelta
+    epoch: int  # an epoch,  used for the calculation of the tps & size_evolution
+    active_connections: int
+    idle: int
+    idle_in_transaction: int
+    idle_in_transaction_aborted: int
+    total: int
+    waiting: int
+    max_connections: int
+    autovacuum_workers: int
+    autovacuum_max_workers: int
+    logical_replication_workers: int
+    parallel_workers: int
+    max_logical_replication_workers: int
+    max_parallel_workers: int
+    max_worker_processes: int
+    max_wal_senders: int
+    max_replication_slots: int
+    wal_senders: int
+    wal_receivers: int
+    replication_slots: int
+    temp_files: int
+    temp_bytes: int
+    # Computed in data.pg_get_server_information() or here
+    size_evolution: float
+    tps: int
+    insert_per_second: int
+    update_per_second: int
+    delete_per_second: int
+    tuples_returned_per_second: int
+
+    @property
+    def worker_processes(self) -> int:
+        if self.parallel_workers is None and self.logical_replication_workers is None:
+            return None
+        else:
+            return (0 if self.parallel_workers is None else self.parallel_workers) + (
+                0
+                if self.logical_replication_workers is None
+                else self.logical_replication_workers
+            )
```

Comment:
There's something wrong here or in fields definition above because most (all?) fields involved here are not declared as `Optional[int]` type so conditions like `self.parallel_workers is None` will never be `True`.

Maybe fields type should be update to, e.g., `Optional[int]`?

Then the property apparently returns `Optional[int]` as well.

Reply:
The typing is wrong and it's true for a good bunch of them.

Reply:
I changed : 
* several attributes to Optional[int].
* the output type for the method.

Added a query file for pre 9.0 versions when I realised max_wal_senders was optional because back then there was no replication. Theses version are untested (I only have builds from 9.0 onwards) but the breakage was obvious so I fixed it.

Reply:
Is it better ?

Reply:
Yes, better; thanks!